timber-junit/src/main/java/net/lachlanmckee/timberjunit/TimberTestRule.java

```diff
@@ -157,71 +198,108 @@ public void evaluate() throws Throwable {
     /**
      * A Timber tree that logs to the Java System.out rather than using the Android logger.
      */
-    private static class JUnitTimberTree extends Timber.DebugTree {
+    private static final class BufferedJUnitTimberTree extends Timber.DebugTree {
         private final Rules mRules;
+        private final List<String> mLogMessageBuffer;
 
-        JUnitTimberTree(Rules rules) {
+        BufferedJUnitTimberTree(Rules rules) {
             mRules = rules;
+            mLogMessageBuffer = new ArrayList<>();
         }
 
         @Override
         protected void log(int priority, String tag, String message, Throwable t) {
-            // Avoid logging if the priority is too low.
-            if (priority < mRules.mMinPriority) {
+            String logMessage = createLogMessage(mRules, priority, tag, message, t);
+            if (logMessage == null) {
                 return;
             }
 
-            // Obtain the correct log type prefix.
-            final char type;
-            switch (priority) {
-                case Log.VERBOSE:
-                    type = 'V';
-                    break;
-
-                case Log.DEBUG:
-                    type = 'D';
-                    break;
-
-                case Log.INFO:
-                    type = 'I';
-                    break;
-
-                case Log.WARN:
-                    type = 'W';
-                    break;
-
-                case Log.ERROR:
-                default:
-                    type = 'E';
-                    break;
-            }
-
-            StringBuilder logBuilder = new StringBuilder();
+            if (mRules.mOnlyLogWhenTestFails) {
+                mLogMessageBuffer.add(logMessage);
 
-            if (mRules.mShowTimestamp) {
-                logBuilder
-                        .append(THREAD_LOCAL_FORMAT.get().format(System.currentTimeMillis()))
-                        .append(" ");
+            } else {
+                System.out.println(logMessage);
             }
+        }
 
-            if (mRules.mShowThread) {
-                Thread thread = Thread.currentThread();
-                logBuilder
-                        .append(thread.getId())
-                        .append("/")
-                        .append(thread.getName())
-                        .append(" ");
+        /**
+         * Flushes all the previously stored log messages.
+         */
+        private void flushLogs() {
+            Iterator<String> iterator = mLogMessageBuffer.iterator();
+            while (iterator.hasNext()) {
+                System.out.println(iterator.next());
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Replace this use of System.out or System.err by a logger. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)